pkg/reconciler/common/prune.go

```diff
@@ -21,75 +21,199 @@ import (
 	"fmt"
 	"os"
 	"regexp"
+	"strconv"
 	"strings"
 
 	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
 	batchv1 "k8s.io/api/batch/v1"
-	v1beta1 "k8s.io/api/batch/v1beta1"
+	"k8s.io/api/batch/v1beta1"
 	corev1 "k8s.io/api/core/v1"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"knative.dev/pkg/logging"
 )
 
 const (
-	tektonSA         = "tekton-pipelines-controller"
-	CronName         = "resource-pruner"
-	JobsTKNImageName = "IMAGE_JOB_PRUNER_TKN"
-	ownerAPIVer      = "operator.tekton.dev/v1alpha1"
-	ownerKind        = "TektonConfig"
+	tektonSA                    = "tekton-pipelines-controller"
+	CronName                    = "resource-pruner"
+	JobsTKNImageName            = "IMAGE_JOB_PRUNER_TKN"
+	ownerAPIVer                 = "operator.tekton.dev/v1alpha1"
+	ownerKind                   = "TektonConfig"
+	pruneSkip                   = "operator.tekton.dev/prune.skip"
+	pruneResources              = "operator.tekton.dev/prune.resources"
+	pruneResourcesPipelineNames = "operator.tekton.dev/prune.resources.pipelineNames"
+	pruneResourcesTaskNames     = "operator.tekton.dev/prune.resources.taskNames"
+	pruneKeep                   = "operator.tekton.dev/prune.keep"
+	pruneKeepSince              = "operator.tekton.dev/prune.keepsince"
```

Comment:
Our goal is to make `keep` and ` keepsince` mutually exclusive.
in the TektonConfig CR prune spec, we can make use of webhook validation to make sure that only one of the 2 is set.

To ensure, similar check occurs in namespace lables, we have to break the prune "strategy" and "value" into 2 labels.


```suggestion
	pruneStrategy                   = "operator.tekton.dev/prune-strategy"
	pruneKeep                         = operator.tekton.dev/prune.keep
	pruneKeepSince               = "operator.tekton.dev/prune.keepsince"
	
```

Our user will have to set either `keep` or `keepSince` as the value for `operator.tekton.dev/prune-strategy` .

the the controller will honor only `operator.tekton.dev/prune.keep` if `operator.tekton.dev/prune-strategy == keep` 

and
the the controller will honor only `operator.tekton.dev/prune.keepSince` if `operator.tekton.dev/prune-strategy == keepSince`

Reply:
Yes, looks good. Will update it.